predicators/pretrained_model_interface.py

```diff
@@ -0,0 +1,247 @@
+# which we use to separate prompt and completion in the cache. The reason to
+# do it this way, rather than saving the prompt and responses separately,
+# is that we want it to be easy to browse the cache as text files.
+_CACHE_SEP = "\n####$$$###$$$####$$$$###$$$####$$$###$$$###\n"
+
+
+class PretrainedLargeModel(abc.ABC):
+    """A pretrained large vision or language model."""
+
+    @abc.abstractmethod
+    def get_id(self) -> str:
+        """Get a string identifier for this model.
+
+        This identifier should include sufficient information so that
+        querying the same model with the same prompt and same identifier
+        should yield the same result (assuming temperature 0).
+        """
+        raise NotImplementedError("Override me!")
+
+    @abc.abstractmethod
+    def _sample_completions(self,
+                            prompt: str,
+                            imgs: Optional[List[PIL.Image.Image]],
+                            temperature: float,
+                            seed: int,
+                            stop_token: Optional[str] = None,
+                            num_completions: int = 1) -> List[str]:
+        """This is the main method that subclasses must implement.
+
+        This helper method is called by sample_completions(), which
+        caches the prompts and responses to disk.
+        """
+        raise NotImplementedError("Override me!")
+
+    def sample_completions(self,
+                           prompt: str,
+                           imgs: Optional[List[PIL.Image.Image]],
+                           temperature: float,
+                           seed: int,
+                           stop_token: Optional[str] = None,
+                           num_completions: int = 1) -> List[str]:
+        """Sample one or more completions from a prompt.
+
+        Higher temperatures will increase the variance in the responses.
+        The seed may not be used and the results may therefore not be
+        reproducible for models where we only have access through an API
+        that does not expose the ability to set a random seed. Responses
+        are saved to disk.
+        """
+        # Set up the cache file.
+        assert _CACHE_SEP not in prompt
+        os.makedirs(CFG.pretrained_model_prompt_cache_dir, exist_ok=True)
+        model_id = self.get_id()
+        prompt_id = hash(prompt)
+        config_id = f"{temperature}_{seed}_{num_completions}_" + \
+                f"{stop_token}"
+        # If the temperature is 0, the seed does not matter.
+        if temperature == 0.0:
+            config_id = f"most_likely_{num_completions}_{stop_token}"
+        cache_foldername = f"{model_id}_{config_id}_{prompt_id}"
+        if imgs is not None:
+            # We also need to hash all the images in the prompt.
+            img_hash_list: List[str] = []
+            for img in imgs:
+                img_hash_list.append(str(imagehash.phash(img)))
+            imgs_id = "".join(img_hash_list)
```

Comment:
is it possible that this folder name will get so long that we'll have issues?

Reply:
there might be ways to combine the hashes into one

Reply:
yes - good point. It seems like I accidentally auto-merged this PR, but let me think + do something about this!

Reply:
actually, upon looking into this a bit, I think let's leave this alone until we actually encounter this issue in-the-wild. I think there are definitely ways to resolve it. I'll make a note for future reference!